### PR TITLE
Adding enhanced error handling for invalid column names in coordinate sources

### LINKED ISSUE #670: Enhance error handling of coordinate source with invalid column names
When we have a coordinate source with invalid column names we don't get a useful error message but a `NullPointerException`. We should be able to give more sensible feedback by first checking if these columns exist or throwing a more useful exception.

```
java.lang.NullPointerException: Cannot invoke "String.concat(String)" because the return value of "java.util.Map.get(Object)" is null
	at edu.ie3.datamodel.io.source.csv.CsvIdCoordinateSource.lambda$buildStreamWithFieldsToAttributesMap$2(CsvIdCoordinateSource.java:123)
	at edu.ie3.datamodel.utils.validation.ValidationUtils.lambda$distinctByKey$5(ValidationUtils.java:245)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:178)
	at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1707)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:960)
	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:934)
	at java.base/java.util.stream.AbstractTask.compute(AbstractTask.java:327)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:754)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:686)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateParallel(ReduceOps.java:927)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at edu.ie3.datamodel.io.source.csv.CsvDataSource.distinctRowsWithLog(CsvDataSource.java:454)
	at edu.ie3.datamodel.io.source.csv.CsvIdCoordinateSource.buildStreamWithFieldsToAttributesMap(CsvIdCoordinateSource.java:124)
	at edu.ie3.datamodel.io.source.csv.CsvIdCoordinateSource.setupIdToCoordinateMap(CsvIdCoordinateSource.java:52)
	at edu.ie3.datamodel.io.source.csv.CsvIdCoordinateSource.<init>(CsvIdCoordinateSource.java:42)
	at edu.ie3.simona.service.weather.WeatherSource$.$anonfun$checkCoordinateSource$3(WeatherSource.scala:460)
	at edu.ie3.simona.service.weather.WeatherSourceWrapper$.apply(WeatherSourceWrapper.scala:321)
	at edu.ie3.simona.service.weather.WeatherSource$.$anonfun$checkConfig$6(WeatherSource.scala:390)
	at edu.ie3.simona.service.weather.WeatherSource$.apply(WeatherSource.scala:295)
	at edu.ie3.simona.service.weather.WeatherService.init(WeatherService.scala:136)
	at edu.ie3.simona.service.SimonaService$$anonfun$uninitialized$1.applyOrElse(SimonaService.scala:64)
	at akka.actor.Actor.aroundReceive(Actor.scala:537)
	at akka.actor.Actor.aroundReceive$(Actor.scala:535)
	at edu.ie3.simona.service.SimonaService.aroundReceive(SimonaService.scala:36)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:579)
	at akka.actor.ActorCell.invoke(ActorCell.scala:547)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:270)
	at akka.dispatch.Mailbox.run(Mailbox.scala:231)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:243)
```

### PR DESCRIPTION
**Pull request by staudtMarius**
Resolves #670

### CODE REVIEW

File: src/main/java/edu/ie3/datamodel/io/source/csv/CsvIdCoordinateSource.java [hunk 1]
```diff
@@ -159,6 +159,9 @@ protected Stream<Map<String, String>> buildStreamWithFieldsToAttributesMap() {
       // returning the original one
       Collection<Map<String, String>> allRows = csvRowFieldValueMapping(reader, headline);
 
+      // checking the column names for each row
+      allRows.forEach(row -> factory.checkForInvalidColumnNames(row.keySet()));
+
```
Comment (sebastian-peter) [inline src/main/java/edu/ie3/datamodel/io/source/csv/CsvIdCoordinateSource.java:164]: I believe it'd be a lot more efficient to check the column headers only once (after line 154) and not repeatedly for every row read.

File: src/main/java/edu/ie3/datamodel/io/factory/timeseries/IdCoordinateFactory.java [hunk 2]
```diff
@@ -28,4 +31,26 @@ protected IdCoordinateFactory() {
 
   /** @return the field id for the coordinate longitude */
   public abstract String getLonField();
+
+  public void checkForInvalidColumnNames(Set<String> columnNames) {
+    List<String> validColumnNames = List.of(getIdField(), getLatField(), getLonField());
+
+    if (!columnNames.containsAll(validColumnNames)) {
+      throw new InvalidColumnNameException(
```
Comment (sebastian-peter) [inline src/main/java/edu/ie3/datamodel/io/factory/timeseries/IdCoordinateFactory.java:39]: Thanks for re-using the existing exception type!

**Review by sebastian-peter — CHANGES_REQUESTED** [review]
Please also check out the SonarQube warnings. To solve them, it could be helpful to check out how `org.slf4j.Logger` can format log messages with supplied objects

**Comment by sebastian-peter** [comment]
Ok I just realized we already have a validation in `Factory#validateParameters`, which is called by `Factory#get(D)`, which is called everytime that any entity is parsed from some source. The problem with IdCoordinateSources is that these validations are called _after_ some other things, like duplicate removal, happened, that already access the data and fail if some column is missing.

Looking at `Factory.get()` as a whole, I'd describe two main issues:
1. validation is sometimes, like in the case of `IdCoordinateSource`, called too late
2. validation is executed for _every single_ data point parsed from `Factory.get()`, which is pretty costly

So I think there's a way to fix both at the same time: Validate column names once during initialization (and early enough).

With regards to this PR: We could merge it to fix the problem in `IdCoordinateSource`, but we then have to remember removing this again once validation is fixed in general.

@staudtMarius @t-ober @danielfeismann

**Comment by staudtMarius** [comment]
Will be resolved in #926.

**Pull request closed without merging**